src/prefect/infrastructure/kubernetes.py

```diff
@@ -590,20 +591,41 @@ def _watch_job(self, job_name: str) -> int:
         # Wait for job to complete
         # if `job_watch_timeout_seconds` is None, no timeout will be enforced
         self.logger.debug(f"Job {job_name!r}: Starting watch for job completion")
-        watch = kubernetes.watch.Watch()
-        with self.get_batch_client() as batch_client:
-            for event in watch.stream(
-                func=batch_client.list_namespaced_job,
-                field_selector=f"metadata.name={job_name}",
-                namespace=self.namespace,
-                timeout_seconds=self.job_watch_timeout_seconds,
+        start_time = pendulum.now("utc")
+        completed = False
+        while not completed:
+            elapsed = (pendulum.now("utc") - start_time).in_seconds()
+            if (
+                self.job_watch_timeout_seconds is not None
+                and elapsed > self.job_watch_timeout_seconds
             ):
-                if event["object"].status.completion_time:
-                    watch.stop()
-                    break
-            else:
-                self.logger.error(f"Job {job_name!r}: Job did not complete.")
+                self.logger.error(f"Job {job_name!r}: Job timed out.")
                 return -1
+            try:
+                watch = kubernetes.watch.Watch()
+                with self.get_batch_client() as batch_client:
+                    remaining_timeout = (
+                        (  # subtract previous watch time
+                            self.job_watch_timeout_seconds - elapsed
+                        )
+                        if self.job_watch_timeout_seconds
+                        else None
+                    )
+
+                    for event in watch.stream(
+                        func=batch_client.list_namespaced_job,
+                        field_selector=f"metadata.name={job_name}",
+                        namespace=self.namespace,
+                        timeout_seconds=remaining_timeout,
+                    ):
+                        if event["object"].status.completion_time:
+                            watch.stop()
+                            completed = True
+            except kubernetes.client.exceptions.ApiException as exc:
+                if exc.status == 410:  # Gone
+                    continue
+                else:
+                    raise exc
```

Comment:
I've seen no evidence of this occurring — do we need this?

If we keep it, we should do `raise` not `raise exc` to avoid adding extra frames.
If the job is gone why would we continue watching it with a `continue`?

Reply:
According to the [k8s docs](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes):

> When the requested watch operations fail because the historical version of that resource is not available, clients must handle the case by recognizing the status code 410 Gone, clearing their local cache, performing a new get or list operation, and starting the watch from the resourceVersion that was returned.

I believe that @zzstoatzz ran into this when doing other work on `prefect-kubernetes` which was the reason that he chose to implement it this way.

Reply:
I do think I missed a step in my initial version of this draft, I think we need to catch the updated resource version when we encounter a 410 with something like

```python
...
                for event in watch.stream(
                    func=batch_client.list_namespaced_job,
                    field_selector=f"metadata.name={job_name}",
                    namespace=self.namespace,
                    timeout_seconds=remaining_timeout_seconds,
                    resource_version=resource_version,
                ):
                    if event["object"].status.completion_time:
                        watch.stop()
                        completed = True
                        break
                    elif event["raw_object"]["code"] == 410:
                        self.logger.info(
                            f"Kubernetes resource version expired for job {job_name!r} - "
                            f"restarting watch from latest resource version: {resource_version}."
                        )
                        resource_version = "0"
                        break
...
```

Reply:
maybe like this?
```python
               resource_version = "0"
               
               ....

                for event in watch.stream(
                    func=batch_client.list_namespaced_job,
                    field_selector=f"metadata.name={job_name}",
                    namespace=self.namespace,
                    timeout_seconds=remaining_timeout_seconds,
                    resource_version=resource_version,
                ):
                    if event["object"].status.completion_time:
                        watch.stop()
                        completed = True
                        break
                    elif event["raw_object"]["code"] == 410:
                        resource_version = event["object"].metadata.resource_version
                        self.logger.info(
                            f"Kubernetes resource version expired for job {job_name!r} - "
                            f"restarting watch from latest resource version: {resource_version}."
                        )
                        break
```

Reply:
Since we're just filtering by job name in the first place, I don't see why we'd get a 410? Since we haven't seen this in the wild yet with flow runs, can we handle this separately once we have a clearer idea of what to do? It seems like this adds complexity and isn't necessary to fix the specific bug we are worried about here.

Reply:
maybe I'm confused about what we're trying to solve here. When reading the issue, it seemed reasonable to me that the fact we weren't handling 410s could be causing the seemingly random 4 hour timeouts / failures described.

I see handling for 410s in popular abstractions on watch ([here](https://github.com/nolar/kopf/blob/7d7a83b8a0b6eb78f30471181a2b77e012b8f492/kopf/_cogs/clients/watching.py#L190) and [here](https://github.com/apache/airflow/blob/2cf1ae30538e109627417e8f0c1650addac3311b/airflow/executors/kubernetes_executor.py#L174)), and resource timeouts are something I encountered when working on k8s collection, but yeah if we want to handle this separately that's fine with me

Reply:
If an exception was being raised it would be in the traceback so we know that wasn't happening for our users at this point. Perhaps it's hidden in the raw object and your sketched solution is necessary, but without more clarity on when and why it occurs I'm quite hesitant to rush a change.

Reply:
okay 👍🏼